Fix: GPIO Stepper Driver

jilaypandya opened the pull request:
- Singularize instantiation of gpio-stepper driver
- Add missing SPDX headers
- Document set_event_callback function in steppers.rst
- Use the correct dtcompatible in release notes
- Rename tests from stepper_api to gpio_steppers
- Add collaborators in stepper drivers

jilaypandya committed:
MAINTAINERS: Add collaborators in stepper drivers

- bjarki-andreasen
- dipakgmx
- fabiobaltieri
- faxe1008

This commit adds dipakgmx as collaborator in stepper drivers

Signed-off-by: Jilay Pandya <jilay.pandya@outlook.com>

kartben commented on doc/hardware/peripherals/stepper.rst:
```diff
@@ -24,6 +24,7 @@ Control Stepper
 - Run continuously with a **constant velocity** in a specific direction until
   a stop is detected using :c:func:`stepper_enable_constant_velocity_mode`.
 - Check if the stepper is **moving** using :c:func:`stepper_is_moving`.
+- Register a **event callback** using :c:func:`stepper_set_event_callback`.
```
```suggestion
- Register an **event callback** using :c:func:`stepper_set_event_callback`.
```

kartben commented on doc/releases/release-notes-4.0.rst:
```diff
@@ -334,7 +334,7 @@ Drivers and Sensors
     stepper motors with :kconfig:option:`CONFIG_STEPPER_SHELL`
   * Added support for ADI TMC5041 (:dtcompatible:`adi,tmc5041`)
-  * Added support for gpio-stepper-controller (:dtcompatible:`gpio-stepper-controller`)
+  * Added support for gpio-stepper-controller (:dtcompatible:`zephyr,gpio-steppers`)
```
fwiw the use of plural is a bit awkward and I am having a hard time finding other bindings using plural form, especially controllers

jilaypandya replied:
Yeah i'll simplify the driver and make it singular

faxe1008 commented on dts/bindings/stepper/zephyr,gpio-stepper.yaml:
```diff
@@ -3,33 +3,29 @@
+  - name: stepper-controller.yaml
+    property-allowlist:
+      - micro-step-res
+      - invert-direction
```
Don't know if this is common practice for drivers generally or if it's fine that if more fields are added they simply go unused?

jilaypandya replied:
you are right atleast in this case all the properties are used. hence have removed `property_allowlist`

jilaypandya commented on drivers/stepper/stepper_shell.c:
```diff
@@ -350,7 +350,7 @@ static int cmd_stepper_set_target_position(const struct shell *sh, size_t argc,
 		return err;
 	}
 
-	err = stepper_set_callback(dev, print_callback, NULL);
```
Passing now `shell_instance` along and also have included the `NULL` pointer check

dipakgmx replied:
You can stick with `NULL`. 

See [can_shell.c](https://github.com/zephyrproject-rtos/zephyr/blob/193eeaef0c04570b80a954b58ccd98075150fb11/drivers/can/can_shell.c#L926), where they the user data is left as `NULL`. (see [can_add_rx_filter()](https://docs.zephyrproject.org/apidoc/latest/group__can__interface.html#gae9dd69a13b960f773ab07bb0bb13b5e9))

jilaypandya replied:
The `shell_instance` is required later in the `print_callback`. I think it's a bit different as to how it is handled in this shell as compared to `can_shell`.

dipakgmx replied:
My bad. Then it makes sense to pass the shell.

dipakgmx commented on drivers/stepper/stepper_shell.c:
```diff
@@ -393,7 +396,7 @@ static int cmd_stepper_enable_constant_velocity_mode(const struct shell *sh, siz
 	}
 
-	err = stepper_set_callback(dev, print_callback, NULL);
+	err = stepper_set_event_callback(dev, print_callback, (void *)sh);
```
Same here. Revert to `NULL`.

bjarki-andreasen commented on drivers/stepper/gpio_stepper_controller.c:
```diff
@@ -77,6 +77,21 @@ static void decrement_coil_charge(const struct device *dev)
+	const struct gpio_stepper_config *config = dev->config;
+
+	for (int i = 0; i < NUM_CONTROL_PINS; i++) {
+		const int err = gpio_pin_set_dt(&config->control_pins[i], GPIO_ACTIVE_LOW);
```
`GPIO_ACTIVE_LOW` is actually 1 `#define GPIO_ACTIVE_LOW         (1 << 0)`, its used to configure the polarity of the pin, you should simply pass 0

bjarki-andreasen reviewed (changes requested):
Regarding the name change of the test from `stepper_api` to `gpio_stepper`, is it no longer possible to use that test with other drivers than `gpio_stepper`?

Regarding collaborators, you can add me as well `bjarki-andreasen` if you wish :)

jilaypandya commented:
> Regarding the name change of the test from `stepper_api` to `gpio_stepper`, is it no longer possible to use that test with other drivers than `gpio_stepper`?

Yeah that's true, In the current state it can only be tested only for the gpio stepper. For other controllers an emul would have to be used in order to test throse controllers. I think the idea of creating the API test in this form was incorrect.

bjarki-andreasen commented:
> > Regarding the name change of the test from `stepper_api` to `gpio_stepper`, is it no longer possible to use that test with other drivers than `gpio_stepper`?
> 
> Yeah that's true, In the current state it can only be tested only for the gpio stepper. For other controllers an emul would have to be used in order to test throse controllers. I think the idea of creating the API test in this form was incorrect.

Well, I know we can't get actual feedback from the mechanical position of the stepper, but we do get feedback in the form of `STEPPER_EVENT_STEPS_COMPLETED`, and we can determine approximately how long it should take to reach the given position. Would the test suite not run perfectly fine on real hardware? its not like we are reading back the actual position from the gpio_stepper driver either, we just wait :)

jilaypandya commented:
> > > Regarding the name change of the test from `stepper_api` to `gpio_stepper`, is it no longer possible to use that test with other drivers than `gpio_stepper`?
> 
> > 
> 
> > Yeah that's true, In the current state it can only be tested only for the gpio stepper. For other controllers an emul would have to be used in order to test throse controllers. I think the idea of creating the API test in this form was incorrect.
> 
> 
> 
> Well, I know we can't get actual feedback from the mechanical position of the stepper, but we do get feedback in the form of `STEPPER_EVENT_STEPS_COMPLETED`, and we can determine approximately how long it should take to reach the given position. Would the test suite not run perfectly fine on real hardware? its not like we are reading back the actual position from the gpio_stepper driver either, we just wait :)

Yeah atleast for gpio stepper this works fine on the hardware as well, I mean since we are just toggling the gpios. And that was also the reason why I renamed it from stepper_api to gpio_stepper. For tmc5041 I have something like an emulator in mind to test it thoroughly.

But yeah this stepper_api test_suite can test only gpio stepper and cannot actually be used for other drivers.

bjarki-andreasen commented:
> > > > Regarding the name change of the test from `stepper_api` to `gpio_stepper`, is it no longer possible to use that test with other drivers than `gpio_stepper`?
> > 
> > > 
> > 
> > > Yeah that's true, In the current state it can only be tested only for the gpio stepper. For other controllers an emul would have to be used in order to test throse controllers. I think the idea of creating the API test in this form was incorrect.
> > 
> > 
> > 
> > Well, I know we can't get actual feedback from the mechanical position of the stepper, but we do get feedback in the form of `STEPPER_EVENT_STEPS_COMPLETED`, and we can determine approximately how long it should take to reach the given position. Would the test suite not run perfectly fine on real hardware? its not like we are reading back the actual position from the gpio_stepper driver either, we just wait :)
> 
> Yeah atleast for gpio stepper this works fine on the hardware as well, I mean since we are just toggling the gpios. And that was also the reason why I renamed it from stepper_api to gpio_stepper. For tmc5041 I have something like an emulator in mind to test it thoroughly.
> 
> But yeah this stepper_api test_suite can test only gpio stepper and cannot actually be used for other drivers.

Even though this test suite is not fully covering the controller, it would run the exact same as the gpio driven driver though, making the test able to run on any hardware right? Just to tezt a base set of features

jilaypandya commented:
> > > > > Regarding the name change of the test from `stepper_api` to `gpio_stepper`, is it no longer possible to use that test with other drivers than `gpio_stepper`?
> > > 
> > > > 
> > > 
> > > > Yeah that's true, In the current state it can only be tested only for the gpio stepper. For other controllers an emul would have to be used in order to test throse controllers. I think the idea of creating the API test in this form was incorrect.
> > > 
> > > 
> > > 
> > > Well, I know we can't get actual feedback from the mechanical position of the stepper, but we do get feedback in the form of `STEPPER_EVENT_STEPS_COMPLETED`, and we can determine approximately how long it should take to reach the given position. Would the test suite not run perfectly fine on real hardware? its not like we are reading back the actual position from the gpio_stepper driver either, we just wait :)
> > 
> > Yeah atleast for gpio stepper this works fine on the hardware as well, I mean since we are just toggling the gpios. And that was also the reason why I renamed it from stepper_api to gpio_stepper. For tmc5041 I have something like an emulator in mind to test it thoroughly.
> > 
> > But yeah this stepper_api test_suite can test only gpio stepper and cannot actually be used for other drivers.
> 
> Even though this test suite is not fully covering the controller, it would run the exact same as the gpio driven driver though, making the test able to run on any hardware right? Just to tezt a base set of features

Ahh yes, now I get what you are pointing to. Thanks 😊. I'll revert it back😊

dkalowsk commented:
That second commit is kind of doing a lot of things here.  Usually I'd flag this as the commit not "doing just one thing".  @kartben what I don't know here is if the change to the MAINTAINERs needs to go through the TSC or not.  

FYI @mmahadevan108

fabiobaltieri reviewed (dismissed):
Yeah commit like this should be split into individual commits, pretty much one one for each point, easier to review, bisect etc... anyway I think it's fine to go with this for this time.

mmahadevan108 commented:
@MaureenHelm, this PR makes changes to the `MAINTAINERS` file. Can you please take a look. Do we need TSC approval to add collaborators?

fabiobaltieri commented:
> Do we need TSC approval to add collaborators?

That is definitely not the case, the recent thing is to add the actual collaborator permissions to users that did not have it previously, adding collaborators is pretty inconsequential.

jilaypandya committed:
drivers: stepper: change gpio-stepper dt-compatible

This commit changes compatible of gpio-stepper in driver

Signed-off-by: Jilay Pandya <jilay.pandya@outlook.com>

jilaypandya committed:
drivers: stepper: fix stepper_set_event_callback c prototype and definition

This commit fixes incorrect c prototype and defintion of
stepper_set_callback to stepper_set_event_callback

Signed-off-by: Jilay Pandya <jilay.pandya@outlook.com>

jilaypandya committed:
drivers: stepper: shell: fix null pointer check

This commit introduces null pointer check in print_callback

Signed-off-by: Jilay Pandya <jilay.pandya@outlook.com>

mmahadevan108 commented:
@jilaypandya , have addressed the changes requested by @kartben

mmahadevan108 commented:
@MaureenHelm can you help review

henrycoding commented:
so when will this request be merged?

fabiobaltieri commented:
> so when will this request be merged?

The project is in hard freeze for the 4.0 release, feature PRs will be merged next week.

mmahadevan108 commented:
After discussing with @jilaypandya , changing the milestone to 4.1
@jilaypandya , please add the `backport-v4.0` label if this is needed for 4.0

nashif commented:
why is this labled with TSC?

MaureenHelm reviewed (changes requested):
> That second commit is kind of doing a lot of things here. Usually I'd flag this as the commit not "doing just one thing".

> Yeah commit like this should be split into individual commits, pretty much one one for each point, easier to review, bisect etc.

The MAINTAINERS commit looks fine, but the second commit needs to be split up.

kartben reviewed (changes requested):
release notes will need to be moved to 4.1, and migraiton guide entry added to indicate the rename of the compatible

jilaypandya committed:
drivers: stepper: gpio: introduce power down coils function

power down coils when gpio stepper is disabled

Signed-off-by: Jilay Pandya <jilay.pandya@outlook.com>

jilaypandya commented:
> release notes will need to be moved to 4.1, and migraiton guide entry added to indicate the rename of the compatible

Yep i have dropped the backport label and the release notes change. This fixes can come in with 4.1

kartben commented:
> > release notes will need to be moved to 4.1, and migraiton guide entry added to indicate the rename of the compatible
> 
> Yep i have dropped the backport label and the release notes change. These fixes can come in with 4.1

Migration guide entry is missing. Renaming the compatible means people will need to update their code

jilaypandya commented:
> > > release notes will need to be moved to 4.1, and migraiton guide entry added to indicate the rename of the compatible
> > 
> > 
> > Yep i have dropped the backport label and the release notes change. These fixes can come in with 4.1
> 
> Migration guide entry is missing. Renaming the compatible means people will need to update their code

can it be a seperate PR?, i will reference it over here.

jilaypandya committed:
doc: migration-guide: 4.1: rename gpio-stepper compatible

add entry in migration guide 4.1 about renaming of compatible from
zephyr,gpio-steppers to zephyr,gpio-stepper

Signed-off-by: Jilay Pandya <jilay.pandya@outlook.com>

jilaypandya commented:
> > > release notes will need to be moved to 4.1, and migraiton guide entry added to indicate the rename of the compatible
> > 
> > 
> > Yep i have dropped the backport label and the release notes change. These fixes can come in with 4.1
> 
> Migration guide entry is missing. Renaming the compatible means people will need to update their code

added
